entity-framework/core/testing/index.md

```diff
@@ -1,117 +1,116 @@
+2. Even when it's possible to involve the real dependency, this may be exceedingly slow: running a large amount of tests against a cloud service may cause tests to take too long. Testing should be part of the developer's everyday workflow, so it's important that tests run quickly.
+3. Executing tests against an external resource may involve isolation issues, where tests interfere with one another. For example, multiple tests running in parallel against a database may modify data and cause each other to fail in various ways. Since unit tests run in-memory and don't use shared resources, they're isolated from each other and don't cause interference.
 
-It is very important to understand that EF Core is not designed to abstract every aspect of the underlying database system.
-Instead, EF Core is a common set of patterns and concepts that can be used with any database system.
-EF Core database providers then layer database-specific behavior and functionality over this common framework.
-This allows each database system to do what it does best while still maintaining commonality, where appropriate, with other database systems.
+However, unit tests do not guarantee that your program works correctly as a whole; while each component may work well in isolation, various issues may arise when integrating them together. For example, a component may execute some query against a database, expecting certain behavior, while the actual database system behaves differently. Such issues are only uncovered in integration testing, making them an important part of any testing strategy. In addition, since unit tests interact with internal components, they must be reworked as the program is refactored and internal APIs change; integration tests, in contrast, are more stable as they test public, external API contracts.
 
-Fundamentally, this means that switching out the database provider will change EF Core behavior and the application can't be expected to function correctly unless it explicitly accounts for any differences in behavior.
-That being said, in many cases doing this will work because there is a high degree of commonality amongst relational databases.
-This is good and bad.
-Good because moving between database systems can be relatively easy.
-Bad because it can give a false sense of security if the application is not fully tested against the new database system.
+## Integration testing
 
-## Approach 1: Production database system
+Because of the above difficulties with integration tests, developers are frequently urged to concentrate on unit tests first, and have a robust unit test suite which they can run frequently on their machines; integration tests, in contrast, are supposed to be executed much less frequently, and in many cases also provide far less coverage. We recommend giving more thought to integration tests, and that that databases are actually far less affected by the above problems as much as people tend to think:
```

Comment:
"We recommend giving more thought to integration tests, and that that databases are actually far less affected by the above problems as much as people tend to think" doesn't parse for me.

Reply:
You're right. Changed to "We recommend giving more thought to integration tests, and suggest that databases may actually be far less affected by the above problems as much as people tend to think", better?